langchain/callbacks/mlflow_callback.py

```diff
@@ -0,0 +1,719 @@
+            run_id=self.run.info.run_id, experiment_id=self.mlf_expid
+        ):
+            self.mlflow.log_metric(key, value)
+
+    def metrics(self, data: dict[str, float], step: Optional[int] = 0):
+        """To log all metrics in the input dict."""
+        with self.mlflow.start_run(
+            run_id=self.run.info.run_id, experiment_id=self.mlf_expid
+        ):
+            self.mlflow.log_metrics(data)
+
+    def jsonf(self, data: dict[str, float], filename: str):
+        """To log the input data as json file artifact."""
+        with self.mlflow.start_run(
+            run_id=self.run.info.run_id, experiment_id=self.mlf_expid
+        ):
+            self.mlflow.log_dict(data, f"{filename}.json")
+
+    def table(self, name: str, df):
+        """To log the input dataframe as a html table"""
+        self.html(df.to_html(), f"table_{name}")
+
+    def html(self, html: str, filename: str):
+        """To log the input html string as html file artifact."""
+        with self.mlflow.start_run(
+            run_id=self.run.info.run_id, experiment_id=self.mlf_expid
+        ):
+            self.mlflow.log_text(html, f"{filename}.html")
+
+    def text(self, text: str, filename: str):
+        """To log the input text as text file artifact."""
+        with self.mlflow.start_run(
+            run_id=self.run.info.run_id, experiment_id=self.mlf_expid
+        ):
+            self.mlflow.log_text(text, f"{filename}.txt")
+
+    def artifact(self, path: str):
+        """To upload the file from given path as artifact."""
+        with self.mlflow.start_run(
+            run_id=self.run.info.run_id, experiment_id=self.mlf_expid
+        ):
+            self.mlflow.log_artifact(path)
+
+    def langchain_artifact(self, chain: Any):
+        with self.mlflow.start_run(
+            run_id=self.run.info.run_id, experiment_id=self.mlf_expid
+        ):
+            self.mlflow.langchain.log_model(chain, "langchain-model")
+
+
+class MlflowCallbackHandler(BaseCallbackHandler):
+    """Callback Handler that logs metrics and artifacts to mlflow server.
+
+    Parameters:
+        name (str): Name of the run.
+        experiment (str): Name of the experiment.
+        tags (str): Tags to be attached for the run.
+        tracking_uri (str): MLflow tracking server uri.
+
+    This handler will utilize the associated callback method called and formats
+    the input of each callback function with metadata regarding the state of LLM run,
+    and adds the response to the list of records for both the {method}_records and
+    action. It then logs the response to mlflow server.
+    """
+
+    def __init__(
+        self,
+        name: Optional[str] = "langchainrun-%",
+        experiment: Optional[str] = "langchain",
+        tags: Optional[Dict] = {},
+        tracking_uri: Optional[str] = None,
+    ) -> None:
+        """Initialize callback handler."""
+        import_pandas()
+        import_textstat()
+        spacy = import_spacy()

```

Comment:
some of the metrics computed here are rather heavy which is why I also looked at hwchase's approach to hide the spacy import/computations behind the vizualize/complexity booleans that can be turned on/off. Consider a similar approach to separate "core tracking" capabilities from "evaluation" which might require more heavy dependencies/computation

https://github.com/akshaya-a/langchain/pull/1/files#diff-a8185aaebbb28c2ebd37436cf39390e57ede177bc301cb7dc897b90f11d3b2bbR199

Reply:
Agree, Will do the required change.